clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/agents/TitusStreamingUpdateAgent.java

```diff
@@ -224,44 +262,60 @@ private void updateJob(Job job) {
       }
     }
 
-    private void updateTask(Task task) {
+    private void updateTask(Task task, Boolean snapshotComplete) {
       String jobId = task.getJobId();
       if (FILTERED_TASK_STATES.contains(task.getStatus().getState())) {
+        tasks.computeIfAbsent(jobId, t -> new HashSet<>()).add(task);
+        changes.incrementAndGet();
+      } else if (FINISHED_TASK_STATES.contains(task.getStatus().getState())) {
         if (tasks.containsKey(jobId)) {
-          tasks.get(jobId).add(task);
+          tasks.get(jobId).remove(task);
+          changes.incrementAndGet();
         } else {
-          tasks.put(jobId, Stream.of(task).collect(Collectors.toCollection(HashSet::new)));
+          if (snapshotComplete) {
+            log.warn("{} jobId {} (task {}) has finished, but not present in current snapshot set",
+              getAgentType(),
+              jobId,
+              task.getId()
+            );
+          }
         }
-        changes.incrementAndGet();
-      } else if (FINISHED_TASK_STATES.contains(task.getStatus().getState())) {
-        tasks.get(jobId).remove(task);
-        changes.incrementAndGet();
       }
     }
 
     /**
-     * Will only write to a cache if the time from the last update exceeds TIME_UPDATE_THRESHHOLD_MS or
-     * the number of changes exceeds ITEMS_CHANGED_THRESHHOLD
+     * Once we persist the first snapshot, we only update the cache if the time from the last update exceeds
+     * TIME_UPDATE_THRESHHOLD_MS or the number of changes exceeds ITEMS_CHANGED_THRESHHOLD.
+     * <p>
+     * TODO: After the full snapshot write, we should support incremental updates to the ProviderCache.
+     * Instead of always calling ProviderCache.putCacheResult (which requires the full result set from
+     * a caching agent, and calls backingStore.mergeAll()), we should call i.e. ProviderCache.putCacheData
+     * (calls backingStore.merge()) and ProviderCache.evictDeletedItems() as needed.
```

Comment:
I hope to follow up on this TODO after the logic changes in this PR are sufficiently tested.